### PR TITLE
New unicode character handling using array data structure.

### PR DESCRIPTION
**Pull request by ktorpi**
JerryScript-DCO-1.0-Signed-off-by: István Kádár ikadar@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/lit/lit-char-helpers.c [hunk 1]
```diff
@@ -14,9 +14,89 @@
...
+    ecma_char_t current = array[middle];
+
+    if (current == c)
+    {
+      return true;
+    }
+
+    if (c < current)
+    {
+      top = middle-1;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: Missing spaces around hypen: top = middle - 1;

File: jerry-core/lit/lit-char-helpers.c [hunk 2]
```diff
@@ -14,9 +14,89 @@
...
+    }
+    else
+    {
+      bottom = middle+1;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: Missing spaces

File: jerry-core/lit/lit-char-helpers.c [hunk 3]
```diff
@@ -14,9 +14,89 @@
...
+ *         false - otherwise
+ */
+static bool
+search_char_in_interval_array (ecma_char_t c,             /**< code unit */
+                               ecma_char_t *array_sp,     /**< array of interval starting points */
+                               uint8_t *lengths,          /**< array of interval lengths */
+                               size_t size_of_array)      /**< length of the array */
+{
+  int bottom = 0;
+  int top = (int) size_of_array - 1;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: I think size_of_array should be int, or top and bottom as size_t. There is no point in changing the sign.

**Comment by LaszloLango** [comment]
typo in title and commit message: *character

File: jerry-core/lit/lit-char-helpers.c [hunk 4]
```diff
@@ -44,14 +124,10 @@ bool
...
-    return true; \
-  }
-#include "lit-unicode-ranges.inc.h"
-
-  return false;
+  return (c == LIT_CHAR_SP || c == LIT_CHAR_NBSP // most common ones
+         || (c >= unicode_separator_char_interv_sps[0]
+            && c <= unicode_separator_char_interv_sps[0] + unicode_separator_char_interv_lens[0])
+         || search_char_in_char_array (c, unicode_separator_chars, NUM_OF_ELEMENTS (unicode_separator_chars)));
 } /* lit_char_is_space_separator */
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.c]: Please optimize this for ASCII:
if (c <= 127)
{
  return c == ... || c == .. etc
}
else
{
 return ...;
}
or optimizing lit_char_is_white_space is even better

File: tools/print-unicode-ranges.sh [hunk 5]
```diff
@@ -17,124 +17,153 @@
...
 #
-UNICODE_DATA_PATH="$1"
 
-#
-# One of unicode character category names (Lu, Ll, Nl, etc.)
-#
-UNICODE_CHAR_CATEGORY="$2"
+# unicode categories:      Lu Ll Lt Mn Mc Me Nd Nl No Zs Zl Zp Cc Cf Cs Co Lm Lo Pc Pd Ps Pe Pi Pf Po Sm Sc Sk So
+# letter:                  Lu Ll Lt Lm Lo Nl
+# non-leter-indent-part:
```
Comment (zherczeg) [inline tools/print-unicode-ranges.sh]: letTer

File: tools/print-unicode-ranges.sh [hunk 6]
```diff
@@ -17,124 +17,153 @@
...
+   FS=";"; \
+   OFS=";"; \
+   is_in_range = 0; \
+   print_count = 0; \
+ } \
+ \
+ function print_Nl() \
+ { \
+    ++print_count; \
+    if (print_count == 10) { \
```
Comment (zherczeg) [inline tools/print-unicode-ranges.sh]: Please fix jerry coding style. { should be in the next line.

File: tools/print-unicode-ranges.sh [hunk 7]
```diff
@@ -17,124 +17,170 @@
...
+#   punctuation mark:      Mn Mc
+#   connector punctuation: Pc
+# separators:              Zs
 
-UNICODE_CHAR_CATEGORY_UPPER_CASE=`echo $UNICODE_CHAR_CATEGORY | tr '[:lower:]' '[:upper:]'`
+if [ $# -le 4 ]; then
+  echo "useage: print-unicode-ranges.sh <unicode-data-path> <-i y sp|y len|n> <-cat letters|non-let-indent-parts|separators>"
+  echo "  -i:   y sp  - print interval starting points"
+  echo "        y len - print interval lengths"
+  echo "        n     - print individual charachters"
```
Comment (LaszloLango) [inline tools/print-unicode-ranges.sh]: *`characters`

File: jerry-core/lit/lit-char-helpers.c [hunk 8]
```diff
@@ -63,13 +121,21 @@ lit_char_is_space_separator (ecma_char_t c) /**< code unit */
...
+    || c == LIT_CHAR_VTAB
+    || c == LIT_CHAR_FF
+    || c == LIT_CHAR_SP;
+  }
+  else
+  {
+    return c == LIT_CHAR_NBSP
+    || c == LIT_CHAR_BOM
+    || (c >= unicode_separator_char_interv_sps[0]
+     && c <= unicode_separator_char_interv_sps[0] + unicode_separator_char_interv_lens[0])
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: wrong indentation

File: jerry-core/lit/lit-char-helpers.c [hunk 9]
```diff
@@ -63,13 +121,21 @@ lit_char_is_space_separator (ecma_char_t c) /**< code unit */
...
-          || c == LIT_CHAR_SP
-          || c == LIT_CHAR_NBSP
-          || c == LIT_CHAR_BOM
-          || lit_char_is_space_separator (c));
+  if (c <= 127)
+  {
+    return c == LIT_CHAR_TAB
+    || c == LIT_CHAR_VTAB
+    || c == LIT_CHAR_FF
+    || c == LIT_CHAR_SP;
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: wrong indentation

File: jerry-core/lit/lit-char-helpers.c [hunk 10]
```diff
@@ -82,9 +148,9 @@ bool
 lit_char_is_line_terminator (ecma_char_t c) /**< code unit */
 {
   return (c == LIT_CHAR_LF
-          || c == LIT_CHAR_CR
-          || c == LIT_CHAR_LS
-          || c == LIT_CHAR_PS);
+  || c == LIT_CHAR_CR
+  || c == LIT_CHAR_LS
+  || c == LIT_CHAR_PS);
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: wrong indentation

File: jerry-core/lit/lit-char-helpers.c [hunk 11]
```diff
@@ -181,82 +205,21 @@ lit_char_is_unicode_letter (ecma_char_t c) /**< code unit */
...
-  if (c >= (range_begin) && c <= (range_end)) \
-  { \
-    return true; \
+  else
+  {
+    return search_char_in_interval_array (c, unicode_non_letter_ident_part_interv_sps,
+                                             unicode_non_letter_ident_part_interv_lens,
+                                             NUM_OF_ELEMENTS (unicode_non_letter_ident_part_interv_sps))
+    || search_char_in_char_array (c, unicode_non_letter_ident_part_chars,
+                                          NUM_OF_ELEMENTS (unicode_non_letter_ident_part_chars));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: wrong indentation

**Comment by LaszloLango** [comment]
@ktorpi, what are the benefits of this patch? Could you measure it on RPi2?

File: jerry-core/lit/lit-char-helpers.c (hunk 9 above)
Comment (ktorpi) [inline jerry-core/lit/lit-char-helpers.c]: If I indent these lines I get error messages like this when running precommit:
> lit-char-helpers.c:127: error: Indentation: 12 -> 4. Line: '            || c == LIT_CHAR_VTAB'
>lit-char-helpers.c:128: error: Indentation: 12 -> 4. Line: '            || c == LIT_CHAR_FF'
>lit-char-helpers.c:129: error: Indentation: 12 -> 4. Line: '            || c == LIT_CHAR_SP;'
What could be the problem?

Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: Are you using parenthesis like in the code before?

**Comment by ktorpi** [comment]
@LaszloLango, here are the results of the measurements (thanks to @bzsolt):

> master: 3543d0c
branch: 9962403

                      Benchmark |                   Perf(+ is better)
                     3d-cube.js |   2.620s ->   2.600s :  +0.763%  (+-0.000%) : [+]
         access-binary-trees.js |   1.400s ->   1.410s :  -0.714%  (+-0.000%) : [-]
             access-fannkuch.js |   7.948s ->   7.934s :  +0.176%  (+-1.307%) : [≈]
                access-nbody.js |   3.082s ->   3.122s :  -1.298%  (+-1.563%) : [≈]
    bitops-3bit-bits-in-byte.js |   1.738s ->   1.722s :  +0.921%  (+-0.746%) : [+]
         bitops-bits-in-byte.js |   2.652s ->   2.652s :  +0.000%  (+-1.203%) : [≈]
          bitops-bitwise-and.js |   3.690s ->   3.668s :  +0.596%  (+-1.141%) : [≈]
       controlflow-recursive.js |   0.888s ->   0.898s :  -1.126%  (+-1.475%) : [≈]
                  crypto-aes.js |   4.732s ->   4.650s :  +1.733%  (+-1.113%) : [+]
                  crypto-md5.js |  21.546s ->  21.928s :  -1.773%  (+-0.874%) : [-]
                 crypto-sha1.js |   9.738s ->   9.916s :  -1.828%  (+-0.772%) : [-]
           date-format-tofte.js |   2.222s ->   2.208s :  +0.630%  (+-1.577%) : [≈]
           date-format-xparb.js |   1.086s ->   1.066s :  +1.842%  (+-1.455%) : [+]
                 math-cordic.js |   2.982s ->   2.984s :  -0.067%  (+-0.846%) : [≈]
           math-partial-sums.js |   1.656s ->   1.644s :  +0.725%  (+-0.960%) : [≈]
          math-spectral-norm.js |   1.502s ->   1.494s :  +0.533%  (+-0.967%) : [≈]
                string-fasta.js |   4.558s ->   4.428s :  +2.852%  (+-1.500%) : [+]
                Geometric mean: |                 Speed up: 0.241% (+-0.275%) : [≈]

Binary sizes(bytes):
master: 197240
branch: 191980

File: jerry-core/lit/lit-char-helpers.c (hunk 9 above)
Comment (ktorpi) [inline jerry-core/lit/lit-char-helpers.c]: Parentheses have solved the problem, thank you.

File: jerry-core/lit/lit-char-helpers.c [hunk 12]
```diff
@@ -181,82 +205,21 @@ lit_char_is_unicode_letter (ecma_char_t c) /**< code unit */
...
-  /* Pc */
-#define LIT_UNICODE_RANGE_PC(range_begin, range_end) \
-  if (c >= (range_begin) && c <= (range_end)) \
-  { \
-    return true; \
+  else
+  {
+    return (search_char_in_interval_array (c, unicode_non_letter_ident_part_interv_sps,
+                                             unicode_non_letter_ident_part_interv_lens,
+                                             NUM_OF_ELEMENTS (unicode_non_letter_ident_part_interv_sps))
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: Indentation should be like this:

```C++
search_char_in_interval_array (c,
                               unicode_non_letter_ident_part_interv_sps,
                               unicode_non_letter_ident_part_interv_lens,
                               NUM_OF_ELEMENTS (unicode_non_letter_ident_part_interv_sps)
```

File: jerry-core/lit/lit-char-helpers.c [hunk 13]
```diff
@@ -181,82 +205,21 @@ lit_char_is_unicode_letter (ecma_char_t c) /**< code unit */
...
+            || search_char_in_char_array (c, unicode_non_letter_ident_part_chars,
+                                             NUM_OF_ELEMENTS (unicode_non_letter_ident_part_chars)));
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.c]: ditto

**Comment by LaszloLango** [comment]
Nice improvement, LGTM.

File: tools/print-unicode-ranges.sh [hunk 14]
```diff
@@ -17,124 +17,170 @@
...
-               printf fmt1, "", $3; \
-               printf fmt2, ""; \
-               printf fmt3, "", $4; \
-             } \
-             \
-             printf "\n"; \
-           }'
+awk -v desired_category="$CATEGORY" \
+'BEGIN \
+ { \
```
Comment (zherczeg) [inline tools/print-unicode-ranges.sh]: I think this should be aligned to one more space, so the alignment is divisible by 2.

**Comment by zherczeg** [comment]
After fixing the style issues, LGTM

**Pull request merged**